File: QuickSort/Haskell/quicksort.hs
partition :: Int -> [Int] -> ([Int], [Int])
partition _ [] = ([], [])
partition p (t:q)
    |t < p = let
            (a, b) = partition p q
        in
            (t:a, b)
    |otherwise = let
            (a, b) = partition p q
        in
            (a, t:b)

quicksort :: [Int] -> [Int]
quicksort [] = []
quicksort (t:q) = let 
        (a, b) = partition t q
    in
        (quicksort a) ++ (t:quicksort b)

main = print(quicksort [18, 21, 3, 54, 21, 22, 4, 32, 17, 28, 2, 31, 74, 30])

File: algorithms/JavaScript/QuickSort/QuickSort.js
/* eslint-disable require-jsdoc */
function quickSort(items, left, right) {
  let index;

  if (items.length > 1) {
    left = typeof left != 'number' ? 0 : left;
    right = typeof right != 'number' ? items.length - 1 : right;

    index = partition(items, left, right);

    if (left < index - 1) {
      quickSort(items, left, index - 1);
    }

    if (index < right) {
      quickSort(items, index, right);
    }
  }

  return items;
}

module.exports = {quickSort};
